pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +227,274 @@ pub mod pallet {
+
+			Ok(())
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		fn do_start_refund(
+			pool_id: T::PoolId,
+			maybe_check_manager: Option<&AccountIdOf<T>>,
+			max_currencies: u32,
+		) -> Result<u32, DispatchError> {
+			let mut pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(
+				pool_details.bonded_currencies.len() <= max_currencies.saturated_into(),
+				Error::<T>::TooManyCurrencies
+			);
+
+			// refunding can only be triggered on a live pool
+			ensure!(pool_details.state.is_live(), Error::<T>::PoolNotLive);
+
+			if let Some(caller) = maybe_check_manager {
+				ensure!(pool_details.is_manager(caller), Error::<T>::Unauthorized);
+			}
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_id.clone().into());
+			// nothing to distribute
+			ensure!(!total_collateral_issuance.is_zero(), Error::<T>::NothingToRefund);
+
+			let has_holders = pool_details
+				.bonded_currencies
+				.iter()
+				.any(|asset_id| !T::Fungibles::total_issuance(asset_id.clone()).is_zero());
+			// no token holders to refund
+			ensure!(has_holders, Error::<T>::NothingToRefund);
+
+			// move pool state to refunding
+			pool_details.state.start_refund();
+			Pools::<T>::set(&pool_id, Some(pool_details.clone()));
+
+			Self::deposit_event(Event::RefundingStarted { id: pool_id });
+
+			Ok(pool_details.bonded_currencies.len().saturated_into())
+		}
+
+		fn do_start_destroy_pool(
+			pool_id: T::PoolId,
+			maybe_check_manager: Option<&AccountIdOf<T>>,
+			force_skip_refund: bool,
+			max_currencies: u32,
+		) -> Result<u32, DispatchError> {
+			let mut pool_details = Pools::<T>::get(&pool_id).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(
+				pool_details.bonded_currencies.len() <= max_currencies.saturated_into(),
+				Error::<T>::TooManyCurrencies
+			);
+
+			ensure!(
+				pool_details.state.is_live() || pool_details.state.is_refunding(),
+				Error::<T>::PoolNotLive
+			);
+
+			if let Some(caller) = maybe_check_manager {
+				ensure!(pool_details.is_manager(caller), Error::<T>::Unauthorized); // TODO: should this be permissionless if the pool is in refunding state?
+			}
+
+			if !force_skip_refund {
+				let total_collateral_issuance = Self::get_pool_collateral(&pool_id.clone().into());
+				if !total_collateral_issuance.is_zero() {
+					let has_holders = pool_details
+						.bonded_currencies
+						.iter()
+						.any(|asset_id| !T::Fungibles::total_issuance(asset_id.clone()).is_zero());
+					// destruction is only allowed when there are no holders or no collateral to distribute
+					ensure!(!has_holders, Error::<T>::LivePool);
+				}
+			}
+
+			// move to destroying state
+			pool_details.state.start_destroy();
+			Pools::<T>::set(&pool_id, Some(pool_details.clone()));
+
+			// emit this event before the destruction started events are emitted by assets deactivation
+			Self::deposit_event(Event::DestructionStarted { id: pool_id });
```

Comment:
I would prefer to move the event after the for loop. The for loop can still fail.

Reply:
Yes that's right, but it is kinda convention to omit events at the end

Reply:
I'd rather have events in the order of 

1. DestructionStarted (pool_id)
2. DestructionStarted (asset_id_1)
1. DestructionStarted (asset_id_2)
2.  ...

than the other way around, don't you think? Having the Pool's DestructionStarted event first provides the context for the events to come.

Reply:
To be honest, I don't care about the order. No strong opinion on that. Feel free to resolve the conversation. I just wanted to bring up my personal preferences

Reply:
😬